diff-models/core/src/common.ts

```diff
@@ -0,0 +1,200 @@
+type Chain = 'provider' | 'consumer';
+
+type Validator = number;
+
+enum Status {
+  BONDED = 'bonded',
+  UNBONDING = 'unbonding',
+  UNBONDED = 'unbonded',
+}
+
+/**
+ * Represents undelegation logic in the staking module.
+ */
+interface Undelegation {
+  val: Validator;
+  creationHeight: number;
+  completionTime: number;
+  balance: number;
+  initialBalance: number;
+  onHold: boolean;
+  opID: number;
+  expired: boolean;
+}
+
+/**
+ * Represents unbonding validator logic in the staking module.
+ */
+interface Unval {
+  val: Validator;
+  unbondingHeight: number;
+  unbondingTime: number;
+  onHold: boolean;
+  opID: number;
+}
+
+/**
+ * Validator Set Change data structure
+ */
+interface Vsc {
+  vscID: number;
+  updates: Record<Validator, number>;
+  downtimeSlashAcks: number[];
+}
+
+/**
+ * Validator Set Change Maturity notification data structure
+ */
+interface VscMatured {
+  vscID: number;
+}
+
+/**
+ * Consumer Initiated Slash data structure
+ */
+interface Slash {
+  val: Validator;
+  vscID: number;
+  isDowntime: boolean;
+}
+
+type PacketData = Vsc | VscMatured | Slash;
+
+interface Packet {
+  data: PacketData;
+  sendHeight: number;
+}
+
+interface Action {
+  kind: string;
+}
+
+type Delegate = {
+  kind: string;
+  val: Validator;
+  amt: number;
+};
+
+type Undelegate = {
+  kind: string;
+  val: Validator;
+  amt: number;
+};
+
+type ConsumerSlash = {
+  kind: string;
+  val: Validator;
+  infractionHeight: number;
+  isDowntime: boolean;
+};
+
+type UpdateClient = {
+  kind: string;
+  chain: Chain;
+};
+
+type Deliver = {
+  kind: string;
+  chain: Chain;
+  numPackets: number;
+};
+
+type EndAndBeginBlock = {
+  kind: string;
+  chain: Chain;
+};
+
+type InvariantSnapshot = {
+  h: Record<Chain, number>;
+  t: Record<Chain, number>;
+  tokens: number[];
+  undelegationQ: Undelegation[];
+  delegatorTokens: number;
+  power: (number | undefined)[];
+};
+
+/**
+ * Record a snapshot of both chain states at the height and timestamp
+ * of a committed block for a chain.
+ */
+interface CommittedBlock {
+  chain: Chain;
+  invariantSnapshot: InvariantSnapshot;
+}
+
+interface Consequence {
```

Comment:
I'd like to know more about what a Consequence is. Is it just the complete model state at any given time? Is it the parts of the model state that you care about checking, minus some "hidden" state?

Reply:
Good point, this is poorly documented. A consequence is a slice of model state relevant to the action that was last done. It's used for state comparison with the SUT and is helpful for understanding traces when reading through the actions. 
Previously I had much more state here, but it's too much and make the jsons huge and any problems really hard to debug.